domain/relation/service/relation.go

```diff
@@ -501,30 +507,40 @@ func (s *Service) EnterScope(
 	}
 
 	// Enter the unit into the relation scope.
-	err := s.st.EnterScope(ctx, relationUUID, unitName, settings)
+	relUnitUUID, err := s.st.EnterScope(ctx, relationUUID, unitName, settings)
 	if errors.Is(err, relationerrors.RelationUnitAlreadyExists) {
 		return nil
 	} else if err != nil {
 		return errors.Capture(err)
 	}
 
+	defer func() {
+		if err != nil {
+			// If there is any failure attempting to create the subordinate unit,
+			// roll back the relation unit creation. EnterScope may only succeed
+			// if the subordinate unit creation succeeds when the subordinate unit
+			// should be created.
+			backoutErr := s.st.DeleteRelationUnit(ctx, relUnitUUID)
+			if backoutErr != nil {
+				s.logger.Errorf(ctx, "rollback of relation unit on failed subordinate creation: %w", backoutErr)
+			}
+		}
+	}()
```

Comment:
We don't need this, just return the error. The uniter will keep retrying indefinitely until it returns no error.